ListGroups and DescribeGroups protocol

wooyeong opened the pull request:
#769

wooyeong commented:
Oh, I think correlationId and request object should go inside mapValuesSeries before this.queueCallback. I'll fix it.

hyperlink commented on lib/kafkaClient.js:
```diff
@@ -413,6 +413,81 @@ KafkaClient.prototype.getApiVersions = function (broker, cb) {
+      // removing {error: null} before merging
+      results = _.map(results, result => _.omitBy(result, _.isNull));
+
+      callback(null, _.merge({}, ...(_.values(results))));
```
We are still supporting node 4 so you can't use spread operators here.

wooyeong replied:
I'll change it to use `.apply` instead

hyperlink commented on lib/kafkaClient.js:
```diff
@@ -413,6 +413,81 @@ KafkaClient.prototype.getApiVersions = function (broker, cb) {
+      this.queueCallback(broker.socket, correlationId, [
+        protocol.decodeListGroups,
+        cb,
+        this.options.versions.requestTimeout
```
this can be omitted if using the default `requestTimeout` value.

hyperlink commented on lib/kafkaClient.js:
```diff
@@ -413,6 +413,81 @@ KafkaClient.prototype.getApiVersions = function (broker, cb) {
+    return callback(new Error('Client is not ready (getListGroups)'));
+  }
+  const brokers = this.brokerMetadata;
+  async.mapValuesSeries(brokers,
```
Series is not need here since there's no no ordering requirement on the result correct? Could we change this to `mapValuesLimit` instead with a default (configurable) limit of 10.

hyperlink commented on lib/kafkaClient.js:
```diff
@@ -413,6 +413,81 @@ KafkaClient.prototype.getApiVersions = function (broker, cb) {
+  if (!this.ready) {
+    return callback(new Error('Client is not ready (getDescribeGroups)'));
+  }
+  const self = this;
```
Is setting `self` necessary since we're using fat arrows throughout?

hyperlink commented on lib/kafkaClient.js:
```diff
@@ -413,6 +413,81 @@ KafkaClient.prototype.getApiVersions = function (broker, cb) {
+        return;
+      }
+
+      async.mapValuesSeries(results,
```
`mapValuesLimit` ?

hyperlink commented on lib/kafkaClient.js:
```diff
@@ -413,6 +413,81 @@ KafkaClient.prototype.getApiVersions = function (broker, cb) {
+          this.queueCallback(broker.socket, correlationId, [
+            protocol.decodeDescribeGroups,
+            cb,
+            this.options.versions.requestTimeout
```
same comment as above this is optional.

wooyeong commented:
I've changed all mentioned problems, and introduced a new option `options.maxAsyncRequests` which defaults to 10 to feed `mapValuesLimit`.

For test, I think I can't write tests right now because I have at least minimum working code and have  to make service first. I may try in the 1st week of October if it's not done.

hyperlink commented on lib/kafkaClient.js:
```diff
@@ -413,6 +414,70 @@ KafkaClient.prototype.getApiVersions = function (broker, cb) {
+    (brokerMetadata, brokerId, cb) => {
+      const broker = this.brokerForLeader(brokerId);
+      if (!broker || !broker.isConnected()) {
+        return callback(new errors.BrokerNotAvailableError('Broker not available (getListGroups)'));
```
`callback` should be `cb` here.

hyperlink commented on lib/kafkaClient.js:
```diff
@@ -413,6 +414,70 @@ KafkaClient.prototype.getApiVersions = function (broker, cb) {
+    return callback(new Error('Client is not ready (getDescribeGroups)'));
+  }
+
+  async.groupBy(groups,
```
Could we use `groupByLimit` here instead?

hyperlink commented on lib/kafkaClient.js:
```diff
@@ -413,6 +414,70 @@ KafkaClient.prototype.getApiVersions = function (broker, cb) {
+        (groups, coordinator, cb) => {
+          const broker = this.brokerForLeader(coordinator);
+          if (!broker || !broker.isConnected()) {
+            return callback(new errors.BrokerNotAvailableError('Broker not available (getDescribeGroups)'));
```
this should be `cb` here and not that main callback.

hyperlink commented on lib/kafkaClient.js:
```diff
@@ -413,6 +414,70 @@ KafkaClient.prototype.getApiVersions = function (broker, cb) {
+          broker.write(request);
+        },
+        (err, res) => {
+          callback(err || null, res);
```
could we flatten the result so it's keyed by `groupId`?

wooyeong commented:
Changed as you requested. I also wrote new admin wrapper class and README.

For `Admin`, I reject if user provides older `Client`, but I don't know it's recommended way to do so.
